kernel/src/snapshot.rs

```diff
@@ -301,104 +332,131 @@ fn read_last_checkpoint(
     }
 }
 
-/// List all log files after a given checkpoint.
-fn list_log_files_with_checkpoint(
-    cp: &CheckpointMetadata,
-    fs_client: &dyn FileSystemClient,
-    log_root: &Url,
-) -> DeltaResult<(Vec<FileMeta>, Vec<FileMeta>)> {
-    let version_prefix = format!("{:020}", cp.version);
-    let start_from = log_root.join(&version_prefix)?;
-
-    let files = fs_client
-        .list_from(&start_from)?
-        .collect::<Result<Vec<_>, Error>>()?
-        .into_iter()
-        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
-        .filter(|f| version_from_location(&f.location).is_some())
-        .collect::<Vec<_>>();
-
-    let mut commit_files = files
-        .iter()
-        .filter_map(|f| {
-            if LogPath::new(&f.location).is_commit {
-                Some(f.clone())
-            } else {
-                None
-            }
-        })
-        .collect_vec();
-    // NOTE this will sort in reverse order
-    commit_files.sort_unstable_by(|a, b| b.location.cmp(&a.location));
-
-    let checkpoint_files = files
-        .iter()
-        .filter_map(|f| {
-            if LogPath::new(&f.location).is_checkpoint {
-                Some(f.clone())
-            } else {
-                None
-            }
-        })
-        .collect_vec();
-
-    // TODO raise a proper error
-    assert_eq!(checkpoint_files.len(), cp.parts.unwrap_or(1) as usize);
-
-    Ok((commit_files, checkpoint_files))
-}
-
 /// List relevant log files.
 ///
 /// Relevant files are the max checkpoint found and all subsequent commits.
+/// If a checkpoint is provided, it starts listing from there; otherwise, it starts from the beginning.
 fn list_log_files(
     fs_client: &dyn FileSystemClient,
     log_root: &Url,
+    requested_version: Option<Version>,
 ) -> DeltaResult<(Vec<FileMeta>, Vec<FileMeta>)> {
-    let version_prefix = format!("{:020}", 0);
-    let start_from = log_root.join(&version_prefix)?;
+    let start_from = log_root.join(&format!("{:020}", 0))?;
 
-    let mut max_checkpoint_version = -1_i64;
     let mut commit_files = Vec::new();
-    let mut checkpoint_files = Vec::with_capacity(10);
-
+    let mut checkpoint_files = Vec::new();
+    let mut last_checkpoint_version = None;
+    let mut next_checkpoint_version = None;
+
+    // If no requested version, try to get the last checkpoint
+    let last_checkpoint = if requested_version.is_none() {
+        read_last_checkpoint(fs_client, log_root)?
+    } else {
+        None
+    };
+
+    // Iterate through files in the log directory
     for maybe_meta in fs_client.list_from(&start_from)? {
         let meta = maybe_meta?;
         let log_path = LogPath::new(&meta.location);
+        let file_version = log_path.version.map(|v| v as i64).unwrap_or(0);
+
+        // Stop if we've passed the requested version
+        if let Some(req_version) = requested_version {
+            if file_version > req_version as i64 {
```

Comment:
Any idea why the existing code was casting from u64 to i64 (which this new code seems to follow)?

I _suspect_ it's because `0` is signed by default, and that passing `0u64` would eliminate type mismatch issues?